File: app/src/main/java/net/squanchy/support/ContextUnwrapper.java
package net.squanchy.support;

import android.content.Context;
import android.content.ContextWrapper;
import android.support.v7.app.AppCompatActivity;

public final class ContextUnwrapper {

    private ContextUnwrapper() {
        // Non-instantiable utility class
    }

    public static AppCompatActivity unwrapToActivityContext(Context context) {
        if (context == null) {
            throw new IllegalArgumentException("Context cannot be null");
        } else if (context instanceof AppCompatActivity) {
            return (AppCompatActivity) context;
        } else if (context instanceof ContextWrapper) {
            ContextWrapper contextWrapper = (ContextWrapper) context;
            return unwrapToActivityContext(contextWrapper.getBaseContext());
        } else {
            throw new IllegalStateException("Context type not supported: " + context.getClass().getCanonicalName());
        }
    }
}

File: app/src/main/java/net/squanchy/support/ContextUnwrapper.kt
@file:JvmName("ContextUnwrapper")

package net.squanchy.support

import android.content.Context
import android.content.ContextWrapper
import androidx.appcompat.app.AppCompatActivity

tailrec fun Context?.unwrapToActivityContext(): AppCompatActivity =
    when (this) {
        null -> throw IllegalArgumentException("Context must not be null")
        is AppCompatActivity -> this
        is ContextWrapper -> this.baseContext.unwrapToActivityContext()
        else -> throw IllegalArgumentException("Context type not supported: ${this.javaClass.canonicalName}")
    }
